File: sample/kotlin/03_julia_set.kt
import kotlin.math.*


// 03: Sample that outputs a Julia set as a PNG image.
// Implemented with simple loop-centric logic for transpilation compatibility.

fun render_julia(width: Long, height: Long, max_iter: Long, cx: Double, cy: Double): MutableList<Any?> {
    var pixels: MutableList<Any?> = __pytra_as_list(mutableListOf<Any?>())
    var __hoisted_cast_1: Double = __pytra_float(height - 1L)
    var __hoisted_cast_2: Double = __pytra_float(width - 1L)
    var __hoisted_cast_3: Double = __pytra_float(max_iter)
    var y = __pytra_int(0L)
    while (y < __pytra_int(height)) {
        var zy0: Double = ((-1.2) + (2.4 * (__pytra_float(y) / __hoisted_cast_1)))
        var x = __pytra_int(0L)
        while (x < __pytra_int(width)) {
            var zx: Double = ((-1.8) + (3.6 * (__pytra_float(x) / __hoisted_cast_2)))
            var zy: Double = zy0
            var i: Long = 0L
            while ((__pytra_int(i) < __pytra_int(max_iter))) {
                var zx2: Double = (zx * zx)
                var zy2: Double = (zy * zy)
                if ((__pytra_float(zx2 + zy2) > __pytra_float(4.0))) {
                    break
                }
                zy = (((2.0 * zx) * zy) + cy)
                zx = ((zx2 - zy2) + cx)
                i += 1L
            }
            var r: Long = 0L
            var g: Long = 0L
            var b: Long = 0L
            if ((__pytra_int(i) >= __pytra_int(max_iter))) {
                r = 0L
                g = 0L
                b = 0L
            } else {
                var t: Double = (__pytra_float(i) / __hoisted_cast_3)
                r = __pytra_int(255.0 * (0.2 + (0.8 * t)))
                g = __pytra_int(255.0 * (0.1 + (0.9 * (t * t))))
                b = __pytra_int(255.0 * (1.0 - t))
            }
            pixels.add(r)
            pixels.add(g)
            pixels.add(b)
            x += 1L
        }
        y += 1L
    }
    return pixels
}

fun run_julia() {
    var width: Long = 3840L
    var height: Long = 2160L
    var max_iter: Long = 20000L
    var out_path: String = "sample/out/03_julia_set.png"
    var start: Double = __pytra_perf_counter()
    var pixels: MutableList<Any?> = __pytra_as_list(render_julia(width, height, max_iter, (-0.8), 0.156))
    __pytra_write_rgb_png(out_path, width, height, pixels)
    var elapsed: Double = (__pytra_perf_counter() - start)
    __pytra_print("output:", out_path)
    __pytra_print("size:", width, "x", height)
    __pytra_print("max_iter:", max_iter)
    __pytra_print("elapsed_sec:", elapsed)
}

fun main(args: Array<String>) {
    run_julia()
}

File: sample/ruby/03_julia_set.rb
require_relative "py_runtime"


# 03: Sample that outputs a Julia set as a PNG image.
# Implemented with simple loop-centric logic for transpilation compatibility.

def render_julia(width, height, max_iter, cx, cy)
  pixels = __pytra_bytearray()
  __hoisted_cast_1 = __pytra_float((height - 1))
  __hoisted_cast_2 = __pytra_float((width - 1))
  __hoisted_cast_3 = __pytra_float(max_iter)
  __step_0 = __pytra_int(1)
  y = __pytra_int(0)
  while ((__step_0 >= 0 && y < __pytra_int(height)) || (__step_0 < 0 && y > __pytra_int(height)))
    zy0 = ((-1.2) + (2.4 * __pytra_div(y, __hoisted_cast_1)))
    __step_1 = __pytra_int(1)
    x = __pytra_int(0)
    while ((__step_1 >= 0 && x < __pytra_int(width)) || (__step_1 < 0 && x > __pytra_int(width)))
      zx = ((-1.8) + (3.6 * __pytra_div(x, __hoisted_cast_2)))
      zy = zy0
      i = 0
      while __pytra_truthy((i < max_iter))
        zx2 = (zx * zx)
        zy2 = (zy * zy)
        if __pytra_truthy(((zx2 + zy2) > 4.0))
          break
        end
        zy = (((2.0 * zx) * zy) + cy)
        zx = ((zx2 - zy2) + cx)
        i += 1
      end
      r = 0
      g = 0
      b = 0
      if __pytra_truthy((i >= max_iter))
        r = 0
        g = 0
        b = 0
      else
        t = __pytra_div(i, __hoisted_cast_3)
        r = __pytra_int((255.0 * (0.2 + (0.8 * t))))
        g = __pytra_int((255.0 * (0.1 + (0.9 * (t * t)))))
        b = __pytra_int((255.0 * (1.0 - t)))
      end
      pixels.append(r)
      pixels.append(g)
      pixels.append(b)
      x += __step_1
    end
    y += __step_0
  end
  return pixels
end

def run_julia()
  width = 3840
  height = 2160
  max_iter = 20000
  out_path = "sample/out/03_julia_set.png"
  start = __pytra_perf_counter()
  pixels = render_julia(width, height, max_iter, (-0.8), 0.156)
  write_rgb_png(out_path, width, height, pixels)
  elapsed = (__pytra_perf_counter() - start)
  __pytra_print("output:", out_path)
  __pytra_print("size:", width, "x", height)
  __pytra_print("max_iter:", max_iter)
  __pytra_print("elapsed_sec:", elapsed)
end

if __FILE__ == $PROGRAM_NAME
  run_julia()
end
